### PR TITLE
Resolve repository from enterprise GitHub url in http and git protocol scheme

### PR DESCRIPTION
**Pull request by XiongKezhi**
The repository full name resolving code refers to https://github.com/jenkinsci/github-branch-source-plugin/blob/1fd2d6f3e1a607f075aed3895420cd50f06ca8d4/src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubRepositoryInfo.java#L80

context: https://github.com/jenkinsci/github-checks-plugin/issues/49#issuecomment-763910554

### CODE REVIEW

File: src/test/java/io/jenkins/plugins/checks/github/GitSCMChecksContextTest.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+        SCMFacade facade = mock(SCMFacade.class);
+
+        when(facade.findGitSCM(run)).thenReturn(Optional.of(scm));
+        when(facade.getUserRemoteConfig(scm)).thenReturn(config);
+
+        for (String url : new String[]{
+                "git@197.168.2.0:jenkinsci/github-checks-plugin",
+                "git@localhost:jenkinsci/github-checks-plugin",
+                "http://github.com/jenkinsci/github-checks-plugin.git",
+                "https://github.com/jenkinsci/github-checks-plugin.git"
```
Comment (XiongKezhi) [inline src/test/java/io/jenkins/plugins/checks/github/GitSCMChecksContextTest.java:29]: @chri7325 @almostjulian

are there any other enterprise URLs are you using?

**Comment by timja** [comment]
Why not just send a PR to make the class in github branch source public?

File: src/test/java/io/jenkins/plugins/checks/github/GitSCMChecksContextTest.java (hunk 1 above)
Comment (chri7325) [reply src/test/java/io/jenkins/plugins/checks/github/GitSCMChecksContextTest.java:29]: @XiongKezhi currently no but we may add more in my company for different sections so I'm hoping things can be checked in a robust manner.

Comment (almostjulian) [reply src/test/java/io/jenkins/plugins/checks/github/GitSCMChecksContextTest.java:29]: My url is git@<ghcompanyservername.company.local>:<repouser>/<project>.git (all those variables could change) ,  if you're regexing the servername part and using the general pattern, I should be good, but I can't hardcode to localhost or a specific ip or anything like that.

**Comment by XiongKezhi** [comment]
> Why not just send a PR to make the class in github branch source public?

The github branch source only supports https protocols, but git scm also supports git protocols: https://plugins.jenkins.io/git/#using-repositories

File: src/test/java/io/jenkins/plugins/checks/github/GitSCMChecksContextTest.java (hunk 1 above)
Comment (XiongKezhi) [reply src/test/java/io/jenkins/plugins/checks/github/GitSCMChecksContextTest.java:29]: > if you're regexing the servername part and using the general pattern, I should be good

the regex for git protocol is `git@.+:.+\\/.+` and I took the part after after ":", should be fine.
https://github.com/jenkinsci/github-checks-plugin/pull/116/files#diff-14f6b06609ffadf765bcd87e1dfe51a8efa36fa3013fd81cec41aec757a611acR100

**Comment by XiongKezhi** [comment]
I encounter a problem myself while I'm setting a Git SCM project with GitHub App credentials 🤷‍♂️

Is GitHub app credentials indeed compatible with ssh protocol? How did you set up the project with it?
截屏2021-01-29 下午4 32 44

**Comment by chri7325** [comment]
@XiongKezhi you may need to add .ssh keys to your Jenkins master in order for it to be able to evaluate that. The jobs usually run on agents for us but having it on the master allows the UI to be able to evaluate things there. Oh, also make sure you have the known_hosts able to evaluate it. Once you have keys on master, `ssh -T git@github.com`

**Comment by XiongKezhi** [comment]
> @XiongKezhi you may need to add .ssh keys to your Jenkins master in order for it to be able to evaluate that. The jobs usually run on agents for us but having it on the master allows the UI to be able to evaluate things there. Oh, also make sure you have the known_hosts able to evaluate it. Once you have keys on master, `ssh -T git@github.com`

thanks, I forget to add the ssh key to the container; now with this PR, it works well for my repository URL `git@github.com:jenkins-checks-api-test/codingstyle.git`

**Comment by XiongKezhi** [comment]
> The github branch source only supports https protocols, but git scm also supports git protocols: https://plugins.jenkins.io/git/#using-repositories

And it throws an `IllegalArgumentException` which we should not catch when the protocol scheme doesn't match HTTP

File: src/test/java/io/jenkins/plugins/checks/github/GitSCMChecksContextTest.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+import java.util.Optional;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+class GitSCMChecksContextTest {
+    @Test
+    public void shouldGetRepository() {
```
Comment (timja) [inline src/test/java/io/jenkins/plugins/checks/github/GitSCMChecksContextTest.java:16]: could we extract the logic out to pass a string to a method so we don't have to do all this mocking?

**Review by timja — APPROVED** [review]

**Comment by chri7325** [comment]
@XiongKezhi thank you for the fix! I'm now able to see the correct url and have used the webapp to be able to see the checks being modified in our enterprise repo. I now have a slightly different issue. We're using the git plugin to clone the repo using the ghprbSourceBranch and ghprbTargetBranch to clone the target branch and then merge in the source branch. This creates a new SHA which doesn't exist on the origin remote since it is being done locally on the check job. I then get the message

```
[GitHub Checks] Failed Publishing GitHub checks: org.kohsuke.github.GHFileNotFoundException: https://devtopia.esri.com/api/v3/repos/runtime/runtimecore/check-runs {"message":"No commit found for SHA: eb536221eae6d14a4e12d5366ebad0ef71bcc436","documentation_url":"https://developer.github.com/enterprise/2.21/v3/checks/runs/#create-a-check-run"}
```

I could merge in the source branch in a script but using the plugin is a nicer experience. When a clean merge doesn't happen, it doesn't create a new SHA and then the plugin does publish to our Github checks tab. However, if we then try to retry the check, it fails immediately since it does not provide any of the payload variables such as ghprbSourceBranch and ghprbTargetBranch. Is that a known limitation or have I configured something incorrectly?

**Comment by XiongKezhi** [comment]
@chri7325 Sorry for the delay

> However, if we then try to retry the check, it fails immediately since it does not provide any of the payload variables such as ghprbSourceBranch and ghprbTargetBranch.

What do you mean by "retry the check" and provide "payload variables"? And seems you are using the [GitHub Pull Request Builder Plugin](https://github.com/jenkinsci/ghprb-plugin), so does the re-try means you are re-triggering the build by commenting on the PR?

Now, from what I understand of your use-case, this could be a known limitation: required by GitHub, the checks could only be published to the SHA that already exists in the remote repo, so when a clean merge happens locally and hasn't pushed to the remote branch, the checks can't be published.

**Comment by chri7325** [comment]
@XiongKezhi we do use the older Pull request builder as setup at https://devopscube.com/jenkins-build-trigger-github-pull-request. We also then use the Naginator plugin to allow retrying jenkins jobs that fail for various reaons that don't necessarily need to update the PR for. Otherwise, any update to the PR will automatically retrigger the checks. Using the pull request builder, we're able to merge in master into the checks workspace as I'm guessing the older pull request builder may just sign the SHA that comes from the webook payload variables. I was able to move this step into our script and did see that the intergration worked at that point though so there is a workaround for that. I was just curious if there as any plans to further integrate this plugin into the older workflows but I'd understand if that isn't going to be supported as well.

**Comment by XiongKezhi** [comment]
> I was just curious if there as any plans to further integrate this plugin into the older workflows but I'd understand if that isn't going to be supported as well.

it's not in the plan, sorry. The pr request builder is unmaintained and as in its [plugin site](https://plugins.jenkins.io/ghprb/), migrate to the GitHub branch source plugin is recommended.

**Pull request merged**